### PR TITLE
1897: Trigger suite timeout after suite stall.

### PR DESCRIPTION
**Pull request by arjclark**
Addresses #1897 

Not for merge yet, currently breaks (or at very least doesn't finish running) the following tests:
```
20160720T092047Z_cylc_test_special_04-clock-triggered aclark@localhost:7766
20160720T092603Z_cylc_test_events_25-held-not-stalled aclark@localhost:7769
20160720T092833Z_cylc_test_special_05-clock-triggered-utc aclark@localhost:7772
20160720T092125Z_cylc_test_special_06-clock-triggered-iso aclark@localhost:7773
20160720T093329Z_cylc_test_special_08-clock-triggered-0 aclark@localhost:7774
```
The clock triggered tests are basically waiting for some point in the future so lurk around waiting for that and not timing out. The held-not-stalled one never shuts down as a held suite won't time out now as its not stalled (possibly we should address this as a timeout handler condition for a suite that's been held with all tasks in it held starting the timeout clock otherwise you could end up with held suites sitting all over shared servers never shutting down).

@hjoliver - see what you think. If you're happy with the approach in the code change here I'll go ahead and update documentation, tests etc. I could also add in the "fix" dealing held suites proposed above too as part of this change.

### CODE REVIEW

**Comment by hjoliver** [comment]
Unfortunately I think this suggests we need both timeout after stall *and* timeout after prolonged inactivity.

**Comment by arjclark** [comment]
@cylc/core - thoughts on the above?

**Comment by arjclark** [comment]
Discussing with @dpmatthews he and I think the way forward is to update the existing behaviour as proposed in this pull request and follow up with a newer suite-inactivity timeout event as a separate issue (which would do what the existing timeout does). Does this seem agreeable?

**Comment by hjoliver** [comment]
> Does this seem agreeable?

Will it be done before the next release?  Otherwise this could be confusing for anyone using the existing timeout.

Just thinking some more, I forget why we wanted a timeout/delay after before shutting down after a stall, given that a stalled suite is not going to unstall itself between stalling and timing out.  Is it just to allow the operator to intervene manually?   Timeout after lack of activity is different, because the active tasks *might* still finish in time.

**Comment by dpmatthews** [comment]
> Just thinking some more, I forget why we wanted a timeout/delay after before shutting down after a stall, given that a stalled suite is not going to unstall itself between stalling and timing out. Is it just to allow the operator to intervene manually?

Yes. If a task fails most users are going to want the suite to continue running so they can intervene. We added the timeout to deal with the situation where a suite had been forgotten or the user was away for a long period.

> Timeout after lack of activity is different, because the active tasks might still finish in time.

Agreed. This is really just a fail safe. With a well designed suite / well configured site defaults (with polling configured to detect tasks which disappear from the queue without communicating to the suite) it should very rarely happen. The main case I can think of is where a remote host becomes unavailable whilst a task is queued or running.

> Will it be done before the next release? Otherwise this could be confusing for anyone using the existing timeout.

Possibly, although are most users expecting the new behaviour in any case? I agree it would be good to get the new timeout implemented before the next release but I'm not sure it's essential (I think getting the new behaviour implemented is more important because we know it is causing problems).

**Comment by hjoliver** [comment]
Alright, I'm persuaded!

**Comment by arjclark** [comment]
> Will it be done before the next release? Otherwise this could be confusing for anyone using the existing timeout.

> I agree it would be good to get the new timeout implemented before the next release but I'm not sure it's essential (I think getting the new behaviour implemented is more important because we know it is causing problems).

I'll be aiming to get both done in the same release. It may be that it'll be more efficient to put both in the one pull request just so I'm not removing test functionality in this pull request and putting it back in a later one. Leave it with me and I'll see how far I can get/whether its viable to do quickly (pretty much otherwise engaged today and a chunk of tomorrow either way).

**Comment by arjclark** [comment]
Have now put both in the same pull request (made sense as, as per above comment, I'd have had to remove/break tests and then put them back in anyway). Old behaviour is now migrated to the "inactivity" timeout with new behaviour being slotted in to the existing timeout configuration.

@hjoliver - please review 1 - I think I've caught everything but I've a feeling I might have missed something
@benfitzpatrick - please review 2

**Comment by arjclark** [comment]
Rebased.

**Comment by hjoliver** [comment]
@arjclark - sorry for the delay on this.   I'm getting test failures with
```
AttributeError: 'Scheduler' object has no attribute 'suite_timer_active'
```
(seems not to be initialised unless a timer is set)

**Comment by arjclark** [comment]
Will have a dig - probably mucked something up when I had to rebase everything.

**Comment by arjclark** [comment]
@hjoliver - I think this should have fixed it.

File: doc/suiterc.tex [hunk 1]
```diff
@@ -318,9 +319,22 @@ \subsection{[cylc]}
...
+\begin{myitemize}
+    \item {\em type:} ISO 8601 duration/interval representation (e.g.
+ \lstinline=PT5S=, 5 seconds, \lstinline=PT1S=, 1 second) - minimum 0 seconds.
+    \item {\em default:} (none), unless set at the site level.
+\end{myitemize}
+
+\paragraph[inactivity]{[cylc] \textrightarrow [[events]] \textrightarrow inactivity}
+
+If inactivity is set and the inactivity event is handled, the inactivity event
+handler(s) will be called if there is no activity in it for some period
```
Comment (hjoliver) [inline doc/suiterc.tex]: ... if there is no activity in ~~it~~ the suite for some period...

File: lib/cylc/scheduler.py [hunk 2]
```diff
@@ -117,6 +117,7 @@ class Scheduler(object):
     EVENT_STARTUP = 'startup'
     EVENT_SHUTDOWN = 'shutdown'
     EVENT_TIMEOUT = 'timeout'
+    EVENT_INACTIVITY_TIMEOUT = 'inactivity'
```
Comment (hjoliver) [inline lib/cylc/scheduler.py]: I think I'd prefer the event to be 'inactive' rather than 'inactiviity'.  Opinions?

**Comment by hjoliver** [comment]
I'm now getting 3 test failures on this branch:
```
Test Summary Report
-------------------
./tests/startup/02-state-summary.t (Wstat: 0 Tests: 2 Failed: 1)
  Failed test:  2
./tests/reload/17-state-summary.t  (Wstat: 0 Tests: 2 Failed: 1)
  Failed test:  2
./tests/reload/19-graphing-change.t (Wstat: 0 Tests: 13 Failed: 1)
  Failed test:  13
```

All report (suite err file):
```
2016-08-11T01:03:01Z WARNING - suite stalled
Traceback (most recent call last):
  File "/home/vagrant/cylc/cylc.git/lib/cylc/scheduler.py", line 280, in start
    self.run()
  File "/home/vagrant/cylc/cylc.git/lib/cylc/scheduler.py", line 1506, in run
    self.check_suite_stalled()
  File "/home/vagrant/cylc/cylc.git/lib/cylc/scheduler.py", line 1568, in check_suite_stalled
    self.set_suite_timer()
  File "/home/vagrant/cylc/cylc.git/lib/cylc/scheduler.py", line 921, in set_suite_timer
    self._get_events_conf(self.EVENT_TIMEOUT)
TypeError: unsupported operand type(s) for +: 'float' and 'NoneType'
ERROR CAUGHT: cleaning up before exit
```
```

**Comment by arjclark** [comment]
Weird, I'm not seeing any of these. I'll manually have a clear out of the pyc files etc. and see what's going on,

File: lib/cylc/scheduler.py (hunk 2 above)
Comment (arjclark) [inline lib/cylc/scheduler.py]: Don't mind either way myself. Can change it to "inactive" if you want me to?

**Comment by benfitzpatrick** [comment]
> I think I'd prefer the event to be 'inactive' rather than 'inactiviity'. Opinions?

Either is fine by me!

File: lib/cylc/scheduler.py (hunk 2 above)
Comment (hjoliver) [inline lib/cylc/scheduler.py]: Changed my mind - leave it as is.

**Comment by hjoliver** [comment]
> Weird, I'm not seeing any of these.

That is weird.  I've just checked again, same result.  The tests pass on master.  The commit log confirms I really am on your branch.

**Comment by arjclark** [comment]
Caught it - something in my user config is preventing me having these problems.

**Comment by arjclark** [comment]
@hjoliver - should now work :grimacing:

File: tests/events/28-inactivity/suite.rc [hunk 3]
```diff
@@ -0,0 +1,21 @@
+[cylc]
+    UTC mode = True
+    [[events]]
+        inactivity = PT10S
+        abort on inactivity = True
+        timeout handler = cylc stop %(suite)s
```
Comment (benfitzpatrick) [inline tests/events/28-inactivity/suite.rc]: Not `inactivity handler`?

Comment (arjclark) [inline tests/events/28-inactivity/suite.rc]: Nope, testing `abort on inactivity` there rather than timeout.

**Comment by benfitzpatrick** [comment]
Tests pass for me, over to @hjoliver

**Comment by hjoliver** [comment]
Yep, all good here now.

**Pull request merged**